packages/matter.js/src/device/AutoDevice.ts

```diff
@@ -0,0 +1,361 @@
+    ? S & ServerType<R>
+    : T extends []
+    ? {}
+    : T;
+
+/**
+ * Unified state type for a base server and a set of clusters.  This defines
+ * the state type for a device that implements the listed clusters.
+ */
+export type StateType<C extends Constructor, T> =
+    T extends [ClusterInterface<infer S, any, any>]
+    ? InstanceType<C>["state"] & S
+    : T extends [ClusterInterface<infer S, any, any>, ...infer R]
+    ? S & StateType<C, R>
+    : T extends []
+    ? InstanceType<C>["state"]
+    : never;
+
+/**
+ * Methods related to state management.
+ */
+export interface StateManager<State extends BaseState> {
+    /**
+     * Access current attribute values.
+     */
+    get state(): State;
+
+    /**
+     * Update internal state.
+     * 
+     * @param state an object mapping attribute names to new values
+     */
+    set(changes: Partial<State>): void;
+
+    /**
+     * Process a validated write request from the fabric.  The default
+     * implementation applies all writes directly to internal state.
+     * 
+     * Derivatives should override this method to process state changes that
+     * have external dependencies.
+     * 
+     * @param state the attribute values to change
+     */
+    write(changes: Partial<State>): Promise<void>;
+}
+
+type Constructor = new (...args: any[]) => any;
+type UntypedClusterServer = <() => >;
+
+/**
+ * Construction options for AutoDevice.
+ */
+export type AutoDeviceOptions<State> = {
+    state?: Partial<State>;
+    endpointId?: number;
+}
+
+// Automatic wiring support class
+class WiringContext {
+    private serverHandlers: UntypedHandlers = {};
+
+    constructor(
+        // The cluster to wire
+        public cluster: ClusterModel,
+
+        // Hooks invoked at construction time to do perform setup
+        public initializers: Array<(this: AutoDevice, options: AutoDeviceOptions<BaseState>) => void>,
+
+        // Target prototype for device class
+        public deviceProto: any,
+
+        // Target prototype for state class
+        public stateProto: any,
+    ) {}
+
+    // Auto-wire state and device interfaces for a specific cluster
+    wire() {
+        this.cluster.attributes.forEach((attr) => {
+            this.wireAttributeState(attr);
+        });
+        this.wireAttributeEvents();
+
+        this.cluster.commands.forEach((command) => {
+            this.wireCommand(command);
+        });
+        
+        this.cluster.events.forEach((event) => {
+            this.wireEvent(event);
+        });
+
+        this.wireClusterServer();
+    }
+
+    // Implement the state interface
+    private wireAttributeState(attr: AttributeModel) {
+        const cluster = this.cluster;
+
+        // Getter
+        const descriptor = <PropertyDescriptor>{
+            get(this: BaseState) {
+                return this[DEVICE].attributeFor(cluster, attr).get();
+            }
+        };
+
+        // Setter
+        if (attr.fixed) {
```

Comment:
should be "not fixed" or? :-)

Reply:
Yeah of course.  I was planning to start on tests today.  Waited longer than normal because things were changing so fast.